compiler/rustc_mir_build/src/check_tail_calls.rs

```diff
@@ -0,0 +1,382 @@
+}
+
+struct TailCallCkVisitor<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    thir: &'a Thir<'tcx>,
+    param_env: ParamEnv<'tcx>,
+    /// Whatever the currently checked body is one of a closure
+    is_closure: bool,
+    /// The result of the checks, `Err(_)` if there was a problem with some
+    /// tail call, `Ok(())` if all of them were fine.
+    found_errors: Result<(), ErrorGuaranteed>,
+    /// Type of the caller function.
+    caller_ty: Ty<'tcx>,
+}
+
+impl<'tcx> TailCallCkVisitor<'_, 'tcx> {
+    fn check_tail_call(&mut self, call: &Expr<'_>, expr: &Expr<'_>) {
+        if self.is_closure {
+            self.report_in_closure(expr);
+            return;
+        }
+
+        let BodyTy::Fn(caller_sig) = self.thir.body_type
+            else { span_bug!(call.span, "`become` outside of functions should have been disallowed by hit_typeck") };
+
+        let ExprKind::Scope { value, .. } = call.kind
+            else { span_bug!(call.span, "expected scope, found: {call:?}") };
+        let value = &self.thir[value];
+
+        if matches!(
+            value.kind,
+            ExprKind::Binary { .. }
+                | ExprKind::Unary { .. }
+                | ExprKind::AssignOp { .. }
+                | ExprKind::Index { .. }
+        ) {
+            self.report_builtin_op(call, expr);
+            return;
+        }
+
+        let ExprKind::Call { ty, fun, ref args, from_hir_call, fn_span } = value.kind
+            else {
+                self.report_non_call(value, expr);
+                return
+            };
+
+        if !from_hir_call {
+            self.report_op(ty, args, fn_span, expr);
+        }
+
+        // Closures in thir look something akin to
+        // `for<'a> extern "rust-call" fn(&'a [closure@...], ()) -> <[closure@...] as FnOnce<()>>::Output {<[closure@...] as Fn<()>>::call}`
+        // So we have to check for them in this weird way...
+        if let &ty::FnDef(did, substs) = ty.kind() {
+            let parent = self.tcx.parent(did);
+            let fn_ = self.tcx.require_lang_item(LangItem::Fn, Some(expr.span));
+            let fn_once = self.tcx.require_lang_item(LangItem::FnOnce, Some(expr.span));
+            let fn_mut = self.tcx.require_lang_item(LangItem::FnMut, Some(expr.span));
+            if [fn_, fn_once, fn_mut].contains(&parent) {
+                if substs.first().and_then(|arg| arg.as_type()).is_some_and(|t| t.is_closure()) {
+                    self.report_calling_closure(
+                        &self.thir[fun],
+                        substs[1].as_type().unwrap(),
+                        expr,
+                    );
+
+                    // Tail calling is likely to cause unrelated errors (ABI, argument mismatches)
+                    return;
+                }
+            };
+        }
+
+        // Erase regions since tail calls don't care about lifetimes
+        let callee_sig =
+            self.tcx.normalize_erasing_late_bound_regions(self.param_env, ty.fn_sig(self.tcx));
```

Comment:
`fn_sig` here panics on the following snippet:
```rust
fn g() {
    become (&g)();
}
```
(or in general any snippet that involves `&fn()` and such)